src/vm/representations.rs

```diff
@@ -32,6 +29,55 @@ pub struct SymbolicExpression {
     pub span: Span,
 }
 
+macro_rules! guarded_string {
+    ($Name:ident, $Label:literal, $Regex:expr) => {
+        #[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+        pub struct $Name (String);
+        impl TryFrom<String> for $Name {
+            type Error = RuntimeErrorType;
+            fn try_from(value: String) -> Result<Self, Self::Error> {
+                // TODO: use lazy static ?
+                let regex_check = $Regex
+                    .expect("FAIL: Bad static regex.");
+                if regex_check.is_match(&value) {
+                    Ok(Self(value))
+                } else {
+                    Err(RuntimeErrorType::BadNameValue($Label, value).into())
+                }
+            }
+        }
+        
+        impl Deref for $Name {
+            type Target = str;
+            fn deref(&self) -> &Self::Target {
+                &self.0
+            }
+        }
+
+        impl Borrow<str> for $Name {
+            fn borrow(&self) -> &str {
+                &self.0
+            }
+        }
+
+        impl Into<String> for $Name {
+            fn into(self) -> String {
+                self.0
+            }
+        }
+
+        #[cfg(test)]
+        impl From<&'_ str> for $Name {
+            fn from(value: &str) -> Self {
+                Self::try_from(value.to_string()).unwrap()
+            }
+        }
+    }
+}
+
+guarded_string!(ClarityName, "ClarityName", Regex::new("^([a-zA-Z]|[-!?+<>=/*])([a-zA-Z0-9]|[-_!?+<>=/*])*$"));
```

Comment:
Can ClarityNames start with punctuation?

Reply:
Currently, yes -- things like the arithmetic operators are legal names (`+`, `-`, etc.).

However, we could change this so that the arithmetic operators are legal names, but nothing else that starts with punctuation is:

```
(^([a-zA-Z])([a-zA-Z0-9]|[-_!?+<>=/*])*)|^[-+=/*]$|^[<>][=]?$
```

Which I think is a good idea -- will change that.

Reply:
Updated in 74257cdc